test/src/unit-capi-query_2.cc

```diff
@@ -2989,5 +2989,143 @@ TEST_CASE_METHOD(
   tiledb_query_free(&query);
   CHECK(query == nullptr);
 
+  remove_array(array_name);
+}
+
+TEST_CASE_METHOD(
+    Query2Fx, "C API: Test query set config", "[capi][config][query]") {
+  std::string array_name = "query_set_config";
+  remove_array(array_name);
+
+  // Create array
+  int32_t dom[] = {1, 6};
+  int32_t extent = 2;
+  create_array(
+      ctx_,
+      array_name,
+      TILEDB_DENSE,
+      {"d1"},
+      {TILEDB_INT32},
+      {dom},
+      {&extent},
+      {"a"},
+      {TILEDB_INT32},
+      {TILEDB_VAR_NUM},
+      {tiledb::test::Compressor(TILEDB_FILTER_NONE, -1)},
+      TILEDB_ROW_MAJOR,
+      TILEDB_ROW_MAJOR,
+      2,
+      false,
+      false);
+
+  // Open array
+  tiledb_array_t* array = nullptr;
+  int rc = tiledb_array_alloc(ctx_, array_name.c_str(), &array);
+  CHECK(rc == TILEDB_OK);
+  rc = tiledb_array_open(ctx_, array, TILEDB_WRITE);
+  CHECK(rc == TILEDB_OK);
+
+  // Create query
+  tiledb_query_t* query = nullptr;
+  rc = tiledb_query_alloc(ctx_, array, TILEDB_WRITE, &query);
+  CHECK(rc == TILEDB_OK);
+
+  // Create config
+  tiledb_error_t* err;
+  tiledb_config_t* config = nullptr;
+  rc = tiledb_config_alloc(&config, &err);
+  REQUIRE(rc == TILEDB_OK);
+  rc = tiledb_config_set(config, "sm.var_offsets.bitsize", "32", &err);
+  CHECK(rc == TILEDB_OK);
+  rc = tiledb_config_set(config, "sm.var_offsets.extra_element", "true", &err);
+  CHECK(rc == TILEDB_OK);
+  rc = tiledb_config_set(config, "sm.var_offsets.mode", "elements", &err);
+  CHECK(rc == TILEDB_OK);
+
+  // Test setting config
+  rc = tiledb_query_set_config(ctx_, query, config);
+  CHECK(rc == TILEDB_OK);
+
+  // Test modified behavior
+  std::vector<uint32_t> offsets = {0, 1, 2, 4, 7, 9, 10};
+  // even in elements mode, we need to pass offsets size as if uint64
```

Comment:
right, this is ugly, sorry :-( ( @joe-maley FYI)
I guess we never used directly the C API with those "arrow" settings, this ugliness is hidden if using the C++ API.
if you think it's a problem, you can open a ticket and we can track it as part of the known issues with the arrow integration.